Added new gulpfile sample to README file with ES2015 version

jeremenichelli opened the pull request:
I added a new gulpfile sample from gulp 4.0 branch, more simple for new developers, who may not know about sourcemaps or coffee script, in my opinion and a version of it using **ES2015** with side notes about Babel installation requirement. [ issue  #1550 ]

_split view recommended_

jeremenichelli commented:
#1550

travm commented on README.md:
````diff
@@ -23,67 +23,157 @@ For a Getting started guide, API docs, recipes, making a plugin, etc. see the [d
+npm install --save-dev babel-core -babel-preset-es2015
+```
+
+Then create a **.babelrc** file with the rpeset configuration.
````
Small typo here: preset*

travm commented on README.md:
```diff
@@ -23,67 +23,157 @@ For a Getting started guide, API docs, recipes, making a plugin, etc. see the [d
-  // You can use multiple globbing patterns as you would with `gulp.src`
-  // If you are using del 2.0 or above, return its promise
-  return del(['build']);
+  del([ 'assets' ]);
```
I believe you still need the return here.

travm commented on README.md:
```diff
@@ -23,67 +23,157 @@ For a Getting started guide, API docs, recipes, making a plugin, etc. see the [d
-      .pipe(concat('all.min.js'))
-    .pipe(sourcemaps.write())
-    .pipe(gulp.dest('build/js'));
+  return gulp.src(paths.scripts.src)
```
Thoughts on leaving the sourcemaps in this example? Or is it over complicating the example for absolute beginners?

jeremenichelli replied:
As I mentioned above, I don't think it is mandatory to explain how gulp works and what can be achieved with it. But of course if the gulp team thinks it should be there I can reincorporate it.

Another option is to add another example with sourcemaps too.

phated replied:
New versions of vinyl-fs support a sourcemaps option, we should show that.

contra replied:
leave sourcemaps, but use the new sourcemaps syntax in gulp 4

jeremenichelli replied:
I'm not familiar with it, can you answer with a quick example?

travm replied:
`return gulp.src('main.js', { sourcemaps: true })`

jeremenichelli replied:
So, this would be the final result?

```js
function scripts() {
  return gulp.src(paths.scripts, { sourcemaps: true })
    .pipe(coffee())
    .pipe(uglify())
    .pipe(concat('all.min.js'))
     // don't know if this is still needed
     .pipe(sourcemaps.write())
    .pipe(gulp.dest('build/js'));
}
```

travm replied:
Yep, that should do it. You can remove the sourcemaps.write() for that example.

jeremenichelli replied:
is `gulp.sourcemaps` package still required for that to work in gulp 4.0?

travm replied:
Nope, you can remove it.

contra commented on README.md:
```diff
@@ -23,67 +23,157 @@ For a Getting started guide, API docs, recipes, making a plugin, etc. see the [d
-var imagemin = require('gulp-imagemin');
-var sourcemaps = require('gulp-sourcemaps');
-var del = require('del');
+var gulp = require('gulp'),
```
use const

jeremenichelli replied:
This is the **ES5** sample.

contra commented on README.md:
```diff
@@ -23,67 +23,157 @@ For a Getting started guide, API docs, recipes, making a plugin, etc. see the [d
+  less = require('gulp-less'),
+  cssmin = require('gulp-cssmin'),
+  concat = require('gulp-concat'),
+  uglify = require('gulp-uglify'),
```
our coding styleguide doesn't permit var declarations like this

jeremenichelli replied:
OK, so one `var` declaration for each module, right?

phated replied:
yes

contra commented on README.md:
```diff
@@ -23,67 +23,157 @@ For a Getting started guide, API docs, recipes, making a plugin, etc. see the [d
+ * and you can use all packages available on npm, but it must return either a
+ * Promise, a Stream or take a callback and call it
+ */
+function clean() {
```
wouldn't this be a good place for a one line arrow fn?

```js
const clean = () => del(paths.assets)
```

jeremenichelli replied:
YEAP!

phated replied:
@contra I'd prefer these to stay named functions instead of arrow functions assigned to a variable.  Registering named functions by the `Function.name` is a killer feature.

jeremenichelli replied:
It's your call guys :smiley:

phated replied:
maybe we should make a compromise and show clean like `export () => del() as clean` (is that valid syntax) and the others as named functions.

jeremenichelli replied:
We can add a special comment there just for educational purpose indicating that you can do it both ways and avoid possible confusion for the reader.

phated replied:
:+1: perfect

jeremenichelli replied:
Would be good to have a final definition around https://github.com/gulpjs/gulp/pull/1550/files#r54960080, I can get everything in a commit for a final review.

phated replied:
done

contra commented on README.md:
```diff
@@ -23,67 +23,157 @@ For a Getting started guide, API docs, recipes, making a plugin, etc. see the [d
+/* 
+ * Define our tasks using plain functions
+ */
+function styles() {
```
arrow fn with implicit return would be nice

phated replied:
@contra I'd prefer these to stay named functions instead of arrow functions assigned to a variable.  Registering named functions by the `Function.name` is a killer feature.

contra commented on README.md:
```diff
@@ -23,67 +23,157 @@ For a Getting started guide, API docs, recipes, making a plugin, etc. see the [d
+}
+
+function scripts() {
+  return gulp.src(paths.scripts.src)
```
arrow fn with implicit return would be nice

phated replied:
@contra I'd prefer these to stay named functions instead of arrow functions assigned to a variable.  Registering named functions by the `Function.name` is a killer feature.

contra commented on README.md:
```diff
@@ -23,67 +23,157 @@ For a Getting started guide, API docs, recipes, making a plugin, etc. see the [d
+}
+
+function watch() {
+  gulp.watch(paths.scripts.src, scripts);
```
arrow fn with no implicit would be nice

phated replied:
@contra I'd prefer these to stay named functions instead of arrow functions assigned to a variable.  Registering named functions by the `Function.name` is a killer feature.

phated commented on README.md:
```diff
@@ -23,67 +23,157 @@ For a Getting started guide, API docs, recipes, making a plugin, etc. see the [d
+  gulp.watch(paths.styles.src, styles);
 }
+
+gulp.task(clean);
```
these should be using es2015 exports since that is supported now

jeremenichelli replied:
Example, please :pray:

jeremenichelli replied:
Just `export clean`?

phated replied:
https://github.com/gulpjs/gulp/blob/master/docs/recipes/exports-as-tasks.md

jeremenichelli replied:
So, I imagine this is only for the **ES2015** version. Is it necessary to declare it in the export to work or can I export the function itself? Like...

```js
// named task
function build() {
  return gulp.src('src/*.js')
    .pipe(babel())
    .pipe(gulp.dest('lib'));
}

export { build };
```

phated replied:
I don't think that's how exports work.  Exports should only be in the ES2015 example but we probably want to start suggesting people use `module.exports = {}` in ES5 gulpfiles.  @contra thoughts?

phated replied:
Let's do the module.exports

jeremenichelli replied:
To end up the discussion around exporting and regarding the example above with the `export { build }`, what would be the correct ES2015 syntax, considering we will have an `export clean () => ..` statement above.

jeremenichelli replied:
Maybe just adding the `export` word to every named function**?**

phated replied:
@jeremenichelli I think the ES2015 syntax is something like:
```js
export function watch(){ ... }
```
and the ES5 syntax could be:
```js
exports.watch = function watch(){ ... }
```

jeremenichelli replied:
Yeap! Great, I'll have the commit ready for tomorrow so you guys can give it a final review. Thanks for the patience and answering everything.

contra commented on README.md:
```diff
@@ -23,67 +23,157 @@ For a Getting started guide, API docs, recipes, making a plugin, etc. see the [d
+ * Notice that you can specify if tasks can run in parallel or not
+ * using `gulp.series` and `gulp.parallel`
+ */
+gulp.task('build',
```
`gulp.series(clean, gulp.parallel(styles, scripts))` is the correct thing here

phated replied:
use `displayName` instead of a string where possible

contra commented on README.md:
```diff
@@ -23,67 +23,157 @@ For a Getting started guide, API docs, recipes, making a plugin, etc. see the [d
+/*
+ * Expose default task that can be called by just running `gulp` from cli
+ */
+gulp.task('default',
```
no need to wrap in a series

jeremenichelli replied:
Good to know :+1:

contra commented on README.md:
```diff
@@ -23,67 +23,157 @@ For a Getting started guide, API docs, recipes, making a plugin, etc. see the [d
+  gulp.parallel(styles, scripts));
+
+gulp.task('default',
+  gulp.series('build'))
```
this file should be an exact copy of the ES6 one, noticing a few minor differences

jeremenichelli replied:
The only differences I'm noticing is that I'm not repeating the comments in the **ES2015** file. If you're not making reference to that let me know where are the ones you're seeing so I can fix them.

contra commented on README.md:
```diff
@@ -23,67 +23,157 @@ For a Getting started guide, API docs, recipes, making a plugin, etc. see the [d
+ * Register some tasks to expose to the cli
+ */
 gulp.task(clean);
 gulp.task(watch);
```
use the exports pattern instead of gulp.task

phated replied:
https://github.com/gulpjs/gulp/blob/master/docs/recipes/exports-as-tasks.md

jeremenichelli commented on README.md:
```diff
@@ -24,66 +24,154 @@ This file is just a quick sample to give you a taste of what gulp does.
+    .pipe(coffee())
+    .pipe(uglify())
+    .pipe(concat('main.min.js'))
+    .pipe(gulp.dest(paths.scripts.dest));
```
@travm please confirm this is correct when you can

travm replied:
Yep, thats looks correct! :+1:

jeremenichelli commented:
Team, I think I've addressed all the concerns and discussed features. PLease give it another look and let me know if there's something wrong or that I should modify.

jeremenichelli commented:
@travm @phated @contra is this still on board?

phated commented:
@jeremenichelli Sorry about that, I just took on a small contract and am focused on wrapping that up.  I will review again when I have some time.

jeremenichelli commented:
No problem @phated :smiley:

phated commented:
Few things I am going to clean up, but it's looking good

phated commented:
Thanks @jeremenichelli!

jeremenichelli commented:
You're welcome guys, it's a pleasure to collaborate in this!
